src/cmd/load-debug-dump-into-postgres/main.go

```diff
@@ -0,0 +1,290 @@
+	ctx, done := log.SpanContext(ctx, "addFile", zap.String("filename", name))
+	defer done(log.Errorp(&retErr))
+	nfiles.Add(1)
+	ngoro.Add(1)
+	defer ngoro.Add(-1)
+
+	slop := int64(r.Len())
+	defer func() {
+		nbytesout.Add(slop)
+	}()
+
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+		line := s.Bytes()
+		if !utf8.Valid(line) {
+			nbinary.Add(1)
+			continue
+		}
+		nlines.Add(1)
+		nbytesout.Add(int64(len(line)))
+		slop -= int64(len(line)) // account for unknown number of newlines
+
+		hash := blake3.Sum256(line)
+		hashBytes := make([]byte, 32)
+		copy(hashBytes, hash[:])
+
+		var parsed map[string]any
+		var parseError sql.NullString
+		if err := json.Unmarshal(s.Bytes(), &parsed); err != nil {
+			parseError.Valid = true
+			parseError.String = err.Error()
+		} else {
+			njson.Add(1)
+		}
+
+		var ts sql.NullTime
+		for _, key := range []string{"time", "ts", "timestamp"} {
+			t, err := DefaultTimeParser(parsed[key])
+			if err != nil {
+				continue
+			}
+			delete(parsed, key)
+			ts.Valid = true
+			ts.Time = t
+			break
+		}
+		if !ts.Valid && parseError.Valid {
+			// Perhaps this is a pgbouncer log; we will try to parse it.
+			t, f, ok := parsePgBouncerLine(line)
+			if ok {
+				ts.Valid = true
+				ts.Time = t
+				parsed = f
+				parseError.Valid = false
+			}
+		}
+
+		tag, err := db.Exec(ctx, "insert into logs(hash, dumpname, filename, time, line, parsed, parseerror) values ($1, $2, $3, $4, $5, $6, $7) on conflict do nothing", hashBytes, *source, name, ts, line, parsed, parseError)
+		if err != nil {
+			if err := context.Cause(ctx); err != nil {
+				// This is where we exit when C-c is pressed.
+				nerr.Add(1)
+				return err //nolint:wrapcheck
+			}
+			log.Info(ctx, "inserting line failed", zap.Error(err))
+		}
+		switch tag.RowsAffected() {
+		case 0:
+			ndups.Add(1)
+		default:
+			nrows.Add(tag.RowsAffected())
+		}
+	}
+	if err := s.Err(); err != nil {
+		return errors.Wrap(err, "scan")
+	}
+	return nil
+}
+
+func printStats(ctx context.Context) {
+	stats := func(m string) {
+		log.Info(ctx, m, zap.Int64("duplicates", ndups.Load()), zap.Int64("rows", nrows.Load()), zap.Int64("json", njson.Load()), zap.Int64("lines", nlines.Load()), zap.Int64("files", nfiles.Load()), zap.Int64("binary", nbinary.Load()), zap.Int64("bytes_read", nbytesin.Load()), zap.Int64("bytes_processed", nbytesout.Load()), zap.Int64("bytes_in_flight", nbytesin.Load()-nbytesout.Load()), zap.Int64("running", ngoro.Load()), zap.Int64("awaiting_db", nconnacq.Load()))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			stats("final stats")
+			return
+		case <-time.After(time.Second):
+			stats("stats")
+		}
+
+	}
+}
+
+func main() {
+	flag.Parse()
+	done := log.InitBatchLogger("")
+	defer func() {
+		var err error
+		if nerr.Load() > 0 {
+			err = errors.New("dump errored")
+		}
+		done(err)
+	}()
+
+	if *v {
+		log.SetLevel(log.DebugLevel)
+	}
+
+	ctx, c := pctx.Interactive()
```

Comment:
Should you just `defer c()` here? Doesn’t make a ton of different in `main`.

I think conventionally `cancel` might be clearer.

Reply:
defer is too late for the success case (we want to print the "final stats" before exiting)

Reply:
You could close another channel and wait for it before finally exiting, but … that seems crazy, esp. given that this is in `main`. I’m cool with it as-is.